### PR TITLE
close #384

### LINKED ISSUE #384: Heap-use-after-free in FS dir API
This was found when running test-fs.lua: "fs.{open,read,close}dir with more entry".

The uv.fs_readdir(dir, readdir_cb) inside the readdir_cb callback would trigger the crash. Looks like the current FS directory API does not support nested calls.

        local function readdir_cb(err, dirs)
          assert(not err)
          if dirs then
            p(dirs)
            uv.fs_readdir(dir, readdir_cb)  --<-- This would trigger the crash.
          else
            assert(uv.fs_closedir(dir)==true)
          end
        end

=================================================================
==7149==ERROR: AddressSanitizer: heap-use-after-free on address 0x606000003fe0 at pc 0x7f0c83ab04d0 bp 0x7fff181900a0 sp 0x7fff18190098
READ of size 8 at 0x606000003fe0 thread T0
    #0 0x7f0c83ab04cf in uv__fs_readdir_cleanup /home/lua-projects/libuv/src/uv-common.c:689:18
    #1 0x7f0c83ac9300 in uv_fs_req_cleanup /home/lua-projects/libuv/src/unix/fs.c:1870:5
    #2 0x7f0c83ab14c3 in uv__work_done /home/lua-projects/libuv/src/threadpool.c:313:5
    #3 0x7f0c83ab8c6e in uv__async_io /home/lua-projects/libuv/src/unix/async.c:147:5
    #4 0x7f0c83accecc in uv__io_poll /home/lua-projects/libuv/src/unix/linux-core.c:384:11
    #5 0x7f0c83ab9a37 in uv_run /home/lua-projects/libuv/src/unix/core.c:373:5
    #6 0x7f0c83a849ec in luv_run /home/lua-projects/lua-modules/luv/src/loop.c:34:13
    #7 0x5cb016 in lj_BC_FUNCC /home/lua-projects/luajit/src/lj_vm.S:809

0x606000003fe0 is located 0 bytes inside of 56-byte region [0x606000003fe0,0x606000004018)
freed by thread T0 here:
    #0 0x4e7928 in __interceptor_free /home/llvm/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:124
    #1 0x7f0c83aac542 in uv__free /home/lua-projects/libuv/src/uv-common.c:88:3
    #2 0x7f0c83ac3165 in uv__fs_closedir /home/lua-projects/libuv/src/unix/fs.c:524:3
    #3 0x7f0c83ac3165 in uv__fs_work /home/lua-projects/libuv/src/unix/fs.c:1423
    #4 0x7f0c83ac7539 in uv_fs_closedir /home/lua-projects/libuv/src/unix/fs.c:1727:3
    #5 0x7f0c83a9d56b in luv_fs_closedir /home/lua-projects/lua-modules/luv/src/fs.c:819:3
    #6 0x5cb016 in lj_BC_FUNCC /home/lua-projects/luajit/src/lj_vm.S:809

previously allocated by thread T4 here:
    #0 0x4e7d07 in malloc /home/llvm/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146
    #1 0x7f0c83ac154d in uv__fs_opendir /home/lua-projects/libuv/src/unix/fs.c:451:9
    #2 0x7f0c83ac154d in uv__fs_work /home/lua-projects/libuv/src/unix/fs.c:1421
    #3 0x7f0c83ab1ee1 in worker /home/lua-projects/libuv/src/threadpool.c:122:5
    #4 0x7f0c8781e6da in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x76da)

Thread T4 created by T0 here:
    #0 0x434100 in __interceptor_pthread_create /home/llvm/llvm/projects/compiler-rt/lib/asan/asan_interceptors.cc:210
    #1 0x7f0c83ae9ceb in uv_thread_create_ex /home/lua-projects/libuv/src/unix/thread.c:258:9
    #2 0x7f0c83ae9a67 in uv_thread_create /home/lua-projects/libuv/src/unix/thread.c:212:10
    #3 0x7f0c83ab1144 in init_threads /home/lua-projects/libuv/src/threadpool.c:225:9
    #4 0x7f0c83ab1144 in init_once /home/lua-projects/libuv/src/threadpool.c:252
    #5 0x7f0c87826826 in __pthread_once_slow (/lib/x86_64-linux-gnu/libpthread.so.0+0xf826)

SUMMARY: AddressSanitizer: heap-use-after-free /home/lua-projects/libuv/src/uv-common.c:689:18 in uv__fs_readdir_cleanup
Shadow bytes around the buggy address:
  0x0c0c7fff87a0: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0c7fff87b0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
  0x0c0c7fff87c0: fd fd fd fd fd fd fd fd fa fa fa fa fd fd fd fd
  0x0c0c7fff87d0: fd fd fd fd fa fa fa fa fd fd fd fd fd fd fd fd
  0x0c0c7fff87e0: fa fa fa fa fd fd fd fd fd fd fd fd fa fa fa fa
=>0x0c0c7fff87f0: fd fd fd fd fd fd fd fd fa fa fa fa[fd]fd fd fd
  0x0c0c7fff8800: fd fd fd fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8810: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8820: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8830: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c0c7fff8840: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==7149==ABORTING

### PR DESCRIPTION
**Pull request by zhaozg**

### CODE REVIEW

**Comment by zhaozg** [comment]
Done

**Comment by squeek502** [comment]
It might be worth revisiting the `fs_{open,read,close}dir` API's here, as I'm not sure we're handling/binding them correctly.

It's probably a good idea to look at how [node is planning on binding them](https://github.com/nodejs/node/pull/29349), where it has it's own `fs.Dir` type that tracks the `close` state of the `uv_dir_t`, and [in its `gc` it calls `uv_fs_closedir`/`uv_fs_req_cleanup` if it hasn't already been closed](https://github.com/nodejs/node/pull/29349/files#diff-86235cf7ac9263baea1bb33cd80b6239R91).

I think `luv` might want to do the same type of thing: create a new `luv_dir_t` type that keeps track of its `closed` state and has a `__gc` function that calls `uv_fs_closedir`/`uv_fs_req_cleanup` if it hasn't already been closed by the user.

The alternative would be to not have a `__gc` for `uv_dir_t` at all and have the user responsible for avoiding memory leaks by *always* calling `fs_closedir`, but that doesn't seem like the Lua way to do things.

**Comment by zhaozg** [comment]
Done again.

**Comment by zhaozg** [comment]
Done again.

File: src/fs.c [hunk 1]
```diff
@@ -373,8 +371,8 @@ static void luv_fs_cb(uv_fs_t* req) {
     lua_insert(L, -nargs - 1);
     nargs++;
   }
-  luv_fulfill_req(L, (luv_req_t*)req->data, nargs);
-  LUV_FS_CLEANUP_REQ
+  luv_fulfill_req(L, data, nargs);
+  luv_fs_cleanup_req(L, req, data);
```
Comment (squeek502) [inline src/fs.c]: This is not going to work. `uv_fs_req_cleanup` has to be called before `luv_fulfill_req` to avoid the race condition.

**Comment by squeek502** [comment]
Confirmed with valgrind that the use-after-free is fixed. This should be good to go.

**Pull request merged**